feat(api/ui)#50779 : add asset group filtering and asset group views

davidfgcorreia opened the pull request:
Add backend and OpenAPI support for filtering assets by group (`group_pattern`) in /assets endpoint.

Implement multi-asset dependency queries in /ui/dependencies endpoint with node_ids parameter.

Add new frontend pages and components for asset group views, including group graph and sidebar.

Refactor asset list and group list to support group filtering, merging, and deduplication.

Add and update tests for group filtering, group views, and multi-asset dependency queries.

introduced small bug in caching on the asset_grath.

https://youtu.be/ybpRUjEhdoQ

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #50779 

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

davidfgcorreia committed:
feat(api/ui)#50779 : add asset group filtering and asset group views

Add backend and OpenAPI support for filtering assets by group
(`group_pattern`) in /assets endpoint.

Implement multi-asset dependency queries in /ui/dependencies endpoint
with node_ids parameter.

Add new frontend pages and components for asset group views, including
group graph and sidebar.

Refactor asset list and group list to support group filtering, merging,
and deduplication.

Add and update tests for group filtering, group views, and multi-asset
dependency queries.

Co-authored-by: Francisco Núncio <francisco.nuncio@tecnico.ulisboa.pt>

bugraoz93 reviewed (commented):
API side looks good! Could you please fix static checks? It is failing for `Compile / format / lint UI`

guan404ming commented on airflow-core/src/airflow/ui/src/pages/Asset/AssetGroupSidebar.tsx:
```diff
@@ -0,0 +1,94 @@
+        <HStack gap={4} mb={2}>
+          <Box>
+            <Text color="chakra-fg" fontSize="xs">
+              {/* i18n: Producing Tasks label */}Producing Tasks
```
We currently disable literal string in ui due to i18n thus you need to update the translation file like `en/assets.json` or `en/common.json` and use `useTranslation` hook to access them. You could also run `npm run lint` in ui folder to see your lint error.

guan404ming reviewed (commented):
Thanks, this feature is great. Feel free to ask anything.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/dependencies.py:
```diff
@@ -80,9 +95,26 @@ def get_dependencies(session: SessionDep, node_id: str | None = None) -> BaseGra
+                    if edge_tuple not in seen_edges:
+                        all_edges.append(e)
+                        seen_edges.add(edge_tuple)
+            data = {"nodes": all_nodes, "edges": all_edges}
```
I don't think we should do this manually.

`extract_single_connected_component` is actually calling `extract_connected_components` which holds all the connected components of the graph. We could directly use all of those and merge them to avoid computing multiple times `extract_connected_components` (for each node id).

davidfgcorreia replied:
Since extract_connected_components returns all components, should I use it even for single node IDs (by finding the one containing the ID), or should I keep using extract_single_connected_component in that case for simplicity?

pierrejeambrun replied:
Maybe you can rework it a little bit to have:
- extract_connected_components (return all)
- rework and rename extract_single_connected component to instead take a list of node, and extrat the relevant coponent IDs. So you can always use this function everywhere, whether you want a single connected component or multiple of them filtered out.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/dependencies.py:
```diff
@@ -41,12 +40,28 @@
+    ids_to_fetch: list[str] = []
+    # If node_id contains commas, treat as multiple IDs (extra protection)
+    if node_ids:
+        ids_to_fetch = [nid.strip() for nid in node_ids.split(",") if nid.strip()]
```
No, FastAPI handles `node_ids: list[str] | None = None` natively. FastAPI chose exploded form query params, i.e `?node_ids=1&node_ids=2` you will directly receive `node_ids=["1","2"] in your function.

davidfgcorreia replied:
I chose to send the parameters as a string specifically to avoid the exploded format in the URL. node_ids=1&node_ids=2, I send something like ?node_ids=1,2 (or another delimited format).

pierrejeambrun replied:
Please stay consistent with the rest of the API, we don't want some endpoints doing that and some other doing something else, also this will remove the need for you to manually parse that string.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/AssetsGroupedList/AssetsGroupedList.tsx:
```diff
@@ -0,0 +1,297 @@
+      fillRule="evenodd"
+    />
+  </svg>
+);
```
We already do this at different places in the code. (We use chevron icons from react-icon lib instead), you can do something similar.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/AssetsGroupedList/AssetsGroupedList.tsx:
```diff
@@ -0,0 +1,297 @@
+      acc[groupName].push(asset);
+
+      return acc;
+    }, {});
```
I think this logic can holds a few problems, first of the endpoint `useAssetServiceGetAssets` is paginated. It means that you'll only retrieve the first 50 items, and you need a mechanism to display the remaining items (cf tables pagination component).

In addition, there is no guarantee at this point that all assets for a specified group will be on the same page, you could end up with a group missing some assets because those were not returned yet.

I think the easiest approach would be to add a backend ui endpoints to list/retrieve asset groups directly. (I group would list all of it's child asset and we would paginate on groups directly). Maybe there are other approaches possible but I would need to think about it.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/AssetsList/AssetsList.tsx:
```diff
@@ -92,64 +105,111 @@ const createColumns = (translate: (key: string) => string): Array<ColumnDef<Asse
+    limit: pagination.pageSize,
+    offset: pagination.pageIndex * pagination.pageSize,
+    orderBy,
+  });
```
We shouldn't do that manually in the UI. (call the endpoints with two different parameters and then merge results in the UI).

The backend should support this. (You can write a custom filter that will do the search accross both 'asset.name` and `asset.group`, or expand the `search_param_factory` and `_SearchParam` to be able to take a `list[ColumnElement]` as attribute. And perform a search across multiple columns.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useMultiDependencyGraph.ts:
```diff
@@ -0,0 +1,51 @@
+    if (!Boolean(queryData)) {
+      queryClient.setQueryData(key, query.data);
+    }
+  });
```
Nice idea but I think in its current form it could induce some bugs. Setting the cache for a particular node_id, with the result of `useDependenciesServiceGetDependencies` with many node ids. Because the response returned for many node ids will hold more connected components than just the 1 single connected component that we should retrieve from the `nodeId: node.id` filter.

This could therefore display unrelated dependencies when displaying dependencies of a particular node.

pierrejeambrun reviewed (commented):
Nice effort!

I believe there are a few things to update before we can merge but this can be a great feature and addition for Airflow `3.1.0`

pierrejeambrun reviewed (commented):
No idea what is wrong in your setup. You can run pre-commit on main to be sure that everything is in order. It is really hard for me to help you without more context and iterating. Everything should be detailed in the contribution docs where you can find all the steps and troubleshooting help.

davidfgcorreia committed:
feat(api/ui): Add asset group endpoints and improve dependencies API

- Added `/assets/groups` endpoint to return asset groups with their assets and counts.
- Added `AssetGroupResponse` and `AssetGroupCollectionResponse` models to API and OpenAPI schemas.
- Updated backend, OpenAPI YAML, and TypeScript types to support asset group listing and filtering.
- Updated frontend to use new `useAssetServiceGetAssetGroups` hook for grouped asset views and sidebar.
- Refactored asset list and group pages to use new group API and improved i18n.
- Updated dependencies API and frontend to use exploded `node_ids` query parameter (array, not comma-separated).
- Updated backend logic for `/dependencies` to return 404 if any requested node_id is missing.
- Updated and fixed unit tests for dependencies and asset groups to match new API and error handling.
- Improved search and filtering for assets and asset groups in UI.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/dependencies.py:
```diff
@@ -41,12 +42,27 @@
+    node_id: str | None = None,
+    node_ids: list[str] = Query(None, description="List of node ids"),
+) -> BaseGraphResponse:
+    """Dependencies graph. Supports a single node_id or multiple node_ids as exploded query parameters (node_ids=foo&node_ids=bar)."""
```
```suggestion
    """Dependencies graph. Supports a single node_id or multiple node_ids
```

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/dependencies.py:
```diff
@@ -41,12 +42,27 @@
-def get_dependencies(session: SessionDep, node_id: str | None = None) -> BaseGraphResponse:
-    """Dependencies graph."""
+def get_dependencies(
+    node_id: str | None = None,
```
```suggestion
```

We should probably only use `node_ids` even for single node to simplify the API. That's private API so it doens't have to be backward compatible.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/AssetsGroupedList/AssetsGroupedList.tsx:
```diff
@@ -0,0 +1,282 @@
+  <CreateAssetEvent asset={original} withText={false} />
+);
+
+const truncate = (str: string, max = 32) => (str.length > max ? `${str.slice(0, max - 3)}...` : str);
+
+const nameCellRenderer = ({ row: { original } }: { row: { original: AssetResponse } }) => (
+  <NameCell original={original} />
+);
+const lastAssetEventCellRenderer = ({ row: { original } }: { row: { original: AssetResponse } }) => (
+  <LastAssetEventCell original={original} />
+);
+const groupCellRenderer = ({ row: { original } }: { row: { original: AssetResponse } }) => (
+  <GroupCell original={original} />
+);
+const consumingDagsCellRenderer = ({ row: { original } }: { row: { original: AssetResponse } }) => (
+  <ConsumingDagsCell original={original} />
+);
+const producingTasksCellRenderer = ({ row: { original } }: { row: { original: AssetResponse } }) => (
+  <ProducingTasksCell original={original} />
+);
+const triggerCellRenderer = ({ row: { original } }: { row: { original: AssetResponse } }) => (
+  <TriggerCell original={original} />
+);
+
```
Similarly to other columns definition, inline this directly in the columns definition.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/AssetsGroupedList/AssetsGroupedList.tsx:
```diff
@@ -0,0 +1,282 @@
+  const [sort] = sorting;
+  const orderBy = sort ? `${sort.desc ? "-" : ""}${sort.id}` : undefined;
+
+  // Use o novo hook para buscar grupos paginados
```
Comment not in english.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/AssetsList/AssetsList.tsx:
```diff
@@ -92,64 +105,80 @@ const createColumns = (translate: (key: string) => string): Array<ColumnDef<Asse
   const [sort] = sorting;
   const orderBy = sort ? `${sort.desc ? "-" : ""}${sort.id}` : undefined;
 
+  const searchMatchMode = "any"; // "any" para OR, "all" para AND
```
comment not in english.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useMultiDependencyGraph.ts:
```diff
@@ -0,0 +1,53 @@
+} from "openapi/queries";
+import type { BaseGraphResponse } from "openapi/requests/types.gen";
+
+/**
+ * Hook to fetch the dependency graph for a list of asset ids.
+ * Uses the backend endpoint that accepts node_ids as an exploded array (node_ids=foo&node_ids=bar).
+ * Caches the result by ids.
+ */
```
```suggestion

```

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useMultiDependencyGraph.ts:
```diff
@@ -0,0 +1,53 @@
+ * Uses the backend endpoint that accepts node_ids as an exploded array (node_ids=foo&node_ids=bar).
+ * Caches the result by ids.
+ */
+export const useMultiDependencyGraph = (
+  assetIds: Array<string>,
+  options?: Omit<UseQueryOptions<BaseGraphResponse, unknown>, "queryFn" | "queryKey">,
+) => {
+  const queryClient = useQueryClient();
+
+  // Always send nodeIds as an array of asset:... strings
+  const nodeIdsParam = assetIds.map((id) => (id.startsWith("asset:") ? id : `asset:${id}`));
+
+  const query = useDependenciesServiceGetDependencies({ nodeIds: nodeIdsParam }, undefined, options);
+
+  if (query.data) {
+    const key = UseDependenciesServiceGetDependenciesKeyFn({ nodeIds: nodeIdsParam });
+    const queryData = queryClient.getQueryData(key);
+
+    if (!Boolean(queryData)) {
+      queryClient.setQueryData(key, query.data);
+    }
+  }
+
+  return query;
+};
```
That should `useDependencyGraph.ts` we don't need another hook doing the exact same thing.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useMultiDependencyGraph.ts:
```diff
@@ -0,0 +1,53 @@
+
+  const query = useDependenciesServiceGetDependencies({ nodeIds: nodeIdsParam }, undefined, options);
+
+  if (query.data) {
+    const key = UseDependenciesServiceGetDependenciesKeyFn({ nodeIds: nodeIdsParam });
+    const queryData = queryClient.getQueryData(key);
+
+    if (!Boolean(queryData)) {
+      queryClient.setQueryData(key, query.data);
+    }
```
This is done automatically by react query. Since you are setting the cache for `nodeIdsParam` which is one of the query parameters. That is not needed.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/vite-env.d.ts:
```diff
@@ -19,6 +19,7 @@
  */
 
 /// <reference types="vite/client" />
+/// <reference types="vite/types/importMeta.d.ts" />
```
I don't think it's needed anymore. That must have been fixed in other PRs.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -665,7 +665,53 @@ def _transform_ti_states(states: list[str] | None) -> list[TaskInstanceState | N
+class _AssetPatternFilterModel(BaseModel):
+    """Asset Pattern Filter Model with a match mode parameter."""
+
+    name_pattern: str | None = None
+    group_pattern: str | None = None
+    search_match_mode: Literal["any", "all"] | None = None
```
Those should probably be two different params, not just one.

And you can define them easily with the search_param_factory since they are not really doing more.

I don't think you need a custom class

```
QueryDagIdPatternSearch = Annotated[
    _SearchParam, Depends(search_param_factory(DagModel.dag_id, "dag_id_pattern"))
]
QueryDagDisplayNamePatternSearch = Annotated[
    _SearchParam, Depends(search_param_factory(DagModel.dag_display_name, "dag_display_name_pattern"))
]
```

pierrejeambrun replied:
Maybe don't do advanced match mode directly cause we would need to do some refactoring with `_TagsFilter` to have a BaseAdvance match mode filter to extend from. Because there's a lot duplicated.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -125,7 +127,7 @@ def depends(cls, only_active: bool = True) -> OnlyActiveFilter:
     limit: QueryLimit,
     offset: QueryOffset,
-    name_pattern: QueryAssetNamePatternSearch,
+    asset_pattern_filter: QueryAssetPatternFilter,
```
We shouldn't replace like this, but add the `group_pattern_filter` next to it, so we can chose what to filter on.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -206,6 +208,88 @@ def get_assets(
+    )
+    group_rows = session.execute(group_query).all()
+    limit_value = limit.value if limit.value is not None else len(group_rows)
+    # Ordene em Python
```
Please english comments

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -206,6 +208,88 @@ def get_assets(
+        group_rows.sort(key=lambda x: x[0], reverse=True)
+    else:
+        group_rows.sort(key=lambda x: x[0])
+    # Aplique paginação em Python
```
same

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -206,6 +208,88 @@ def get_assets(
+    group_rows = session.execute(group_query).all()
+    limit_value = limit.value if limit.value is not None else len(group_rows)
+    # Ordene em Python
+    if order_by == "count":
+        group_rows.sort(key=lambda x: (x[1], x[0]))
+    elif order_by == "-count":
+        group_rows.sort(key=lambda x: (-x[1], x[0]))
+    elif order_by == "-group":
+        group_rows.sort(key=lambda x: x[0], reverse=True)
+    else:
+        group_rows.sort(key=lambda x: x[0])
```
What is that? Sorting should be on the db side, not on the application side.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -206,6 +208,88 @@ def get_assets(
+    group_counts = {row[0]: row[1] for row in paginated_rows}
+
+    # Query all assets for these groups
+    assets_query = (
+        select(AssetModel)
+        .where(AssetModel.group.in_(groups))
+        .options(
+            subqueryload(AssetModel.scheduled_dags),
+            subqueryload(AssetModel.producing_tasks),
+            subqueryload(AssetModel.consuming_tasks),
+        )
+    )
+    assets = session.scalars(assets_query).all()
```
This will query all assets from the db. Can't happen I think

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:
```diff
@@ -206,6 +208,88 @@ def get_assets(
+    """Get asset groups."""
+    offset_value = offset.value if offset.value is not None else 0
+
+    group_query = (
+        select(AssetModel.group, func.count(AssetModel.id))
+        .where(
+            AssetModel.active.has() if only_active.value else True,
+            *([AssetModel.group == group] if group else []),
+        )
+        .group_by(AssetModel.group)
+    )
```
This will query all assets from the db grouped by asset group, but that could be thousands of them, that' not scalable.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/services/ui/dependencies.py:
```diff
@@ -59,20 +66,34 @@ def extract_single_connected_component(
-    nodes = [node for node in nodes if node["id"] in connected_component]
-    edges = [
+    if len(node_ids) == 1:
+        node_id = node_ids[0]
+        filtered_connected_components = [cc for cc in connected_components if node_id in cc]
+        if len(filtered_connected_components) != 1:
+            raise ValueError(
+                f"Unique connected component not found, got {filtered_connected_components} for connected components of node {node_id}, expected only 1 connected component."
+            )
+        connected_component = filtered_connected_components[0]
+        filtered_nodes = [node for node in nodes if node["id"] in connected_component]
+        filtered_edges = [
+            edge
+            for edge in edges
+            if (edge["source_id"] in connected_component and edge["target_id"] in connected_component)
+        ]
+        return {"nodes": filtered_nodes, "edges": filtered_edges}
+    # Multiple node_ids: merge all relevant components
+    relevant_components = [cc for cc in connected_components if any(nid in cc for nid in node_ids)]
+    # NEW: Check that all node_ids are present in the merged set
+    if not relevant_components:
+        raise ValueError(f"No connected component found for node(s) {node_ids}.")
+    merged_node_ids = set().union(*relevant_components)
+    missing = [nid for nid in node_ids if nid not in merged_node_ids]
+    if missing:
+        raise ValueError(f"No connected component found for node(s) {missing}.")
+    filtered_nodes = [node for node in nodes if node["id"] in merged_node_ids]
+    filtered_edges = [
         edge
         for edge in edges
-        if (edge["source_id"] in connected_component and edge["target_id"] in connected_component)
+        if (edge["source_id"] in merged_node_ids and edge["target_id"] in merged_node_ids)
```
Needs refactoring, that's really hard to read .

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Asset/AssetGroupSidebar.tsx:
```diff
@@ -0,0 +1,97 @@
+          <Heading size="md">{groupName}</Heading>
+        </HStack>
+        <Text color="chakra-fg" fontSize="sm" mb={2}>
+          <b>{translate("group", { ns: "assets" })}</b>
+          <br />
```
use chakra.p, chakra.br, etc.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Asset/AssetGroupSidebar.tsx:
```diff
@@ -0,0 +1,97 @@
+        ) : (
+          events.map((ev) => (
+            <Box bg="chakra-subtle-bg-hover" borderRadius="md" key={ev.id} mb={2} p={2}>
+              <Text color="chakra-fg-subtle" fontSize="xs">
+                {ev.timestamp}
+              </Text>
```
We probably shouldn't display timestamp as text, check other location of the app for `Time` component.

pierrejeambrun reviewed (commented):
Thanks for addressing the previous comments.

I think the PR needs refinement, refactoring, simplifications before we can merge this, I think it still remains an important effort there.

That would be a great addition to `3.1.0`
